### PR TITLE
Dev/mbrl

### PR DESCRIPTION
**Pull request by qiaoting159753**
Add a model-based reinforcement learning module to the repository. 

Algorithm: Dyna, Mismatching No More (Dyna and GAN). 

1. In the algorithm, add a mbrl folder for mbrl algorithms. Now, it is all based on SAC. 
2. Add MLP dynamic and reward prediction. Add ensemble dynamic. Add an Integrated ensemble to train reward prediction and dynamic prediction together.    
3. For memory buffer, add a function to sample next_action and next_reward for training reward prediction. 
4. For memory buffer, add a function to compute mean and std for normalizing states. Training with normalized states can significantly improve accuracy. 
5. Add functions in the helper for normalizing states (the existing one normalizes actions and is not usable).
6. create 'createDyna' function to create the agent with all parts. 
7. add configurations.

### CODE REVIEW

**Comment by qiaoting159753** [comment]
This is the first pull request for MBRL. The following changes will be made: 

Part 1. Networks: 
(1). Add single MLP dynamic prediction and reward prediction. 
(2). Add an Integrated (dynamic + reward) network to train them together. 

Part 2. DYNA algorithm:
(1). Use the SAC algorithm as the actor-critic framework, plus the dyna algorithm. 
(2). Dyna algorithm produces synthetic data to reduce sample complexity.
(3). The generated data is directly fed to update the agent.  
(4). The actions that generate synthetic data are randomly selected (Empirically better). 

(5). Add configurations for the DYNA algorithm. 

Future features: 
(1). Normalization of states.

File: cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+import copy
+import logging
+import os
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+
```
Comment (beardyFace) [inline cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py:15]: rename - DynaSAC, don't use _ in class names

File: cares_reinforcement_learning/networks/World_Models/ensemble_integrated.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        ).mean()
+        pred_rewards = self.reward_network.forward(pred_next_state, next_actions)
+        all_loss = F.mse_loss(pred_rewards, next_rewards) + model_loss.mean()
+        # Update
+        self.all_optimizer.zero_grad()
+        all_loss.backward()
+        self.all_optimizer.step()
+
+
+class Ensemble_World_Reward:
```
Comment (beardyFace) [inline cares_reinforcement_learning/networks/World_Models/ensemble_integrated.py:98]: remove _

File: cares_reinforcement_learning/networks/World_Models/ensemble_integrated.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+                observation_size=observation_size,
+                num_actions=num_actions,
+                hidden_size=hidden_size,
+                lr=lr,
+            )
+            for _ in range(self.num_models)
+        ]
+        self.statistics = {}
+
+    def to(self, device):
```
Comment (beardyFace) [inline cares_reinforcement_learning/networks/World_Models/ensemble_integrated.py:132]: if private make _to - same with any other private functions

**Review by beardyFace — CHANGES_REQUESTED** [review]

File: cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+            state_tensor = torch.FloatTensor(state).unsqueeze(0).to(self.device)
+            if evaluation is False:
+                (action, _, _) = self.actor_net.sample(state_tensor)
+            else:
+                (_, _, action) = self.actor_net.sample(state_tensor)
+            action = action.cpu().data.numpy().flatten()
+        self.actor_net.train()
+        return action
+
+    def true_train_policy(self, states, actions, rewards, next_states, dones):
```
Comment (beardyFace) [inline cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py:101]: rename - if this is a private part of the training - then make it "_train_policy:

File: cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        info["q_values_one"] = q_values_one
+        info["q_values_two"] = q_values_two
+        info["q_values_min"] = torch.minimum(q_values_one, q_values_two)
+        info["critic_loss_total"] = critic_loss_total
+        info["critic_loss_one"] = critic_loss_one
+        info["critic_loss_two"] = critic_loss_two
+        info["actor_loss"] = actor_loss
+        return info
+
+    def train_world_model(self, experiences):
```
Comment (beardyFace) [inline cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py:165]: _train_world_model - it is a private function

File: cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+            states=states,
+            actions=actions,
+            rewards=rewards,
+            next_states=next_states,
+            dones=dones,
+        )
+        # # # Step 3 Dyna add more data
+        self.dyna_generate_and_train(next_states=next_states)
+
+    def dyna_generate_and_train(self, next_states):
```
Comment (beardyFace) [inline cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py:238]: private function

**Review by beardyFace — CHANGES_REQUESTED** [review]

File: cares_reinforcement_learning/util/configurations.py [hunk 7]
```diff
@@ -106,12 +106,25 @@ class TD3Config(AlgorithmConfig):
...
     actor_lr: Optional[float] = 3e-4
-    critic_lr: Optional[float] = 3e-3
-
+    critic_lr: Optional[float] = 3e-4
     gamma: Optional[float] = 0.99
     tau: Optional[float] = 0.005
+    memory: Optional[str] = "MemoryBuffer"
+
 
+class MBRL_DYNAConfig(AlgorithmConfig):
```
Comment (beardyFace) [inline cares_reinforcement_learning/util/configurations.py:115]: Remove _

**Review by beardyFace — CHANGES_REQUESTED** [review]

File: cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py (hunk 1 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py:15]: Changed

File: cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py (hunk 4 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py:101]: It will be exposed to the train_loop.

File: cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py (hunk 6 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py:238]: Changed

File: cares_reinforcement_learning/networks/World_Models/ensemble_integrated.py (hunk 2 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/networks/World_Models/ensemble_integrated.py:98]: Changed.

File: cares_reinforcement_learning/util/configurations.py [hunk 8]
```diff
@@ -105,12 +105,25 @@ class TD3Config(AlgorithmConfig):
 class SACConfig(AlgorithmConfig):
```
[the remark this answers is not in the document]
Comment (qiaoting159753) [reply cares_reinforcement_learning/util/configurations.py:106]: The hyperparameters from the main branch are different from the hyperparameters provided in the original paper. 'MemoryBuffer' was used by default, added the memory attribute here to make the configuration complete.

File: cares_reinforcement_learning/networks/World_Models/ensemble_integrated.py (hunk 3 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/networks/World_Models/ensemble_integrated.py]: Changed.

File: cares_reinforcement_learning/util/configurations.py (hunk 7 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/util/configurations.py]: Changed

File: cares_reinforcement_learning/algorithm/mbrl/mbrl_dyna_sac.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        actions = torch.FloatTensor(np.asarray(actions)).to(self.device)
+        rewards = torch.FloatTensor(np.asarray(rewards)).to(self.device).unsqueeze(1)
+        next_states = torch.FloatTensor(np.asarray(next_states)).to(self.device)
+        dones = torch.LongTensor(np.asarray(dones)).to(self.device).unsqueeze(1)
+        assert len(states.shape) >= 2
+        assert len(actions.shape) == 2
+        assert len(rewards.shape) == 2 and rewards.shape[1] == 1
+        assert len(next_states.shape) >= 2
+        # Step 2 train as usual
+        self.true_train_policy(
```
[the remark this answers is not in the document]
Comment (qiaoting159753) [reply cares_reinforcement_learning/algorithm/mbrl/mbrl_dyna_sac.py]: Changed

File: cares_reinforcement_learning/algorithm/mbrl/__init__.py [hunk 10]
```diff
@@ -0,0 +1,4 @@
+from .mbrl_dyna_sac import MBRL_DYNA_SAC
+
+# from .mbrl_steve_sac import MBRL_STEVE_SAC
```
[the remark this answers is not in the document]
Comment (qiaoting159753) [reply cares_reinforcement_learning/algorithm/mbrl/__init__.py]: Changed.

File: cares_reinforcement_learning/networks/World_Models/__init__.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+from cares_reinforcement_learning.networks.World_Models.ensemble_dynamics import (
```
[the remark this answers is not in the document]
Comment (qiaoting159753) [reply cares_reinforcement_learning/networks/World_Models/__init__.py]: Changed.

File: cares_reinforcement_learning/util/plotter.py [hunk 12]
```diff
@@ -51,8 +51,8 @@ def plot_data(
     )
 
     plt.legend(fontsize="15", loc="upper left")
-    fig_manager = plt.get_current_fig_manager()
-    fig_manager.window.setGeometry(100, 100, 800, 650)
+    # fig_manager = plt.get_current_fig_manager()
```
[the remark this answers is not in the document]
Comment (qiaoting159753) [reply cares_reinforcement_learning/util/plotter.py]: Will change it back.

File: cares_reinforcement_learning/util/uncertainty_estimators.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    return multi_loss.unsqueeze(dim=1).detach()
+
+
+def mean_std(mean):
+    total_stds = torch.std(mean, dim=0)
+    total_stds = torch.sum(total_stds, dim=1)
+    return total_stds.detach().item()
+
+
+def sampling(pred_means, pred_vars):
```
[the remark this answers is not in the document]
Comment (qiaoting159753) [reply cares_reinforcement_learning/util/uncertainty_estimators.py]: This part will wait till future pulls.

File: cares_reinforcement_learning/memory/memory_buffer.py [hunk 14]
```diff
@@ -1,27 +1,106 @@
...
+        :return: statistic tuple of the collected transitions.
+        """
+        states = [trans[0] for trans in self.buffer]
+        next_states = [trans[3] for trans in self.buffer]
+
+        states = np.array(states)
+        next_states = np.array(next_states)
+        delta = next_states - states
+
+        ob_mean = np.mean(states, axis=0) + 0.00001
```
[the remark this answers is not in the document]
Comment (qiaoting159753) [reply cares_reinforcement_learning/memory/memory_buffer.py]: This part will wait till future pulls.

**Review by beardyFace — APPROVED** [review]
